src/graphics/text.rs

```diff
@@ -346,6 +349,169 @@ impl Drawable for Text {
     }
 }
 
+// end - ending index of current word within TextLayout.chars (we always
+// start at 0 and will use the previous word's end to figure out the size
+// of the next word)
+// width - width of the given word in pixels (used to determine whether
+// or not we should start a new line)
+struct Word {
+    end: usize,
+    width: f32,
+}
+
+impl Word {
+    fn from_str(
+        text: &str,
+        font_map: &HashMap<char, CharInfo>,
+        mut upper_bound: usize,
+    ) -> Vec<Self> {
+        let mut buffer = Vec::new();
+        for word in text.split(" ") {
+            upper_bound += word.len();
+            buffer.push(Word {
+                end: upper_bound,
+                width: word
+                    .chars()
+                    .map(|c| font_map.get(&c).unwrap_or(&font_map[&'?']).advance_width)
+                    .sum(),
+            })
+        }
+        buffer
+    }
+}
+
+#[derive(Debug)]
+pub struct LayoutCharInfo {
+    coords: Box2<f32>,
+    color: Color,
+    c: char,
+}
+
+pub struct TextLayout {
+    chars: Vec<LayoutCharInfo>,
+    words: Vec<Word>,
+    font_atlas: FontAtlas,
+    cursor: Point2<f32>,
+    space_width: f32,
+}
+
+impl TextLayout {
+    pub fn new(font_atlas: FontAtlas) -> Self {
+        let space_width = font_atlas.font_map[&' '].advance_width;
+        TextLayout {
+            font_atlas: font_atlas,
+            chars: Vec::new(),
+            words: Vec::new(),
+            cursor: Point2::new(0., 0.),
+            space_width: space_width,
+        }
+    }
+
+    pub fn push_str(&mut self, text: &str, colors: impl IntoIterator<Item = Color> + Clone) {
+        if let Some(upper_bound) = colors.clone().into_iter().size_hint().1 {
```

Comment:
`size_hint` is not necessarily accurate enough for this check. would be better to create a `text.chars()` iterator outside of the main loop; take it mutably using `.by_ref()` when zipped with `colors`; and then assert that it is empty after iterating through everything to make sure there are no left-behind chars